src/coreclr/jit/importer.cpp

```diff
@@ -12429,12 +12430,23 @@ GenTree* Compiler::impCastClassOrIsInstToTree(
             // Jit can only inline expand the normal CHKCASTCLASS helper.
             canExpandInline = (helper == CORINFO_HELP_CHKCASTCLASS);
         }
-        else
+        else if (helper == CORINFO_HELP_ISINSTANCEOFCLASS)
         {
-            if (helper == CORINFO_HELP_ISINSTANCEOFCLASS)
+            // If the class is exact, the jit can expand the IsInst check inline.
+            canExpandInline = isClassExact;
+        }
+
+        if (isClassExact && ((helper == CORINFO_HELP_ISINSTANCEOFARRAY) || (helper == CORINFO_HELP_CHKCASTARRAY)))
+        {
+            // We have "obj isinst/castclass T[]" and in case if T is sealed we can still perform
+            // the fast pMT check (in case if T[] -> obj.GetType is legal in jit-time)
+
+            CORINFO_CLASS_HANDLE elementCls = NO_CLASS_HANDLE;
+            info.compCompHnd->getChildType(pResolvedToken->hClass, &elementCls);
+            if ((elementCls != NO_CLASS_HANDLE) && impIsClassExact(elementCls))
             {
```

Comment:
```suggestion
        if (isClassExact && ((helper == CORINFO_HELP_ISINSTANCEOFARRAY) || (helper == CORINFO_HELP_CHKCASTARRAY)))
        {
```
Why is `isClassExact` not enough?